### ISSUE TITLE
Fargate with spot instances fails with InvalidParameterException

### ISSUE THREAD

**Issue by tracek** [dask/dask-cloudprovider#328]
**What happened**:

Fargate (AWS) with spot instances fails with:

```
InvalidParameterException: An error occurred (InvalidParameterException) when calling the RunTask operation: The specified capacity provider strategy cannot contain a capacity provider that is not associated with the cluster. Associate the capacity provider with the cluster or specify a valid capacity provider and try again.
```

**What you expected to happen**:

Cluster creation succeeds and I get to enjoy spot instances on Fargate

**Minimal Complete Verifiable Example**:

```python
from dask_cloudprovider.aws import FargateCluster

cluster = FargateCluster(fargate_spot=True)
```

**Environment**:

- Dask version: dask cloudprovider: `2021.6.0+16.g876ccba` (master), dask `2021.12.0`
- Python version: 3.8, 3.9
- Operating System: Ubuntu 18.04
- Install method (conda, pip, source): source

<details>

Full stack trace:
```
---------------------------------------------------------------------------
InvalidParameterException                 Traceback (most recent call last)
/tmp/ipykernel_11179/3565260805.py in <module>
----> 1 cluster = FargateCluster(image="<redacted>", fargate_spot=True)

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/dask_cloudprovider/aws/ecs.py in __init__(self, **kwargs)
   1460 
   1461     def __init__(self, **kwargs):
-> 1462         super().__init__(fargate_scheduler=True, fargate_workers=True, **kwargs)
   1463 
   1464 

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/dask_cloudprovider/aws/ecs.py in __init__(self, fargate_scheduler, fargate_workers, fargate_spot, image, scheduler_cpu, scheduler_mem, scheduler_timeout, scheduler_extra_args, scheduler_task_kwargs, scheduler_address, worker_cpu, worker_nthreads, worker_mem, worker_gpu, worker_extra_args, worker_task_kwargs, n_workers, workers_name_start, workers_name_step, cluster_arn, cluster_name_template, execution_role_arn, task_role_arn, task_role_policies, cloudwatch_logs_group, cloudwatch_logs_stream_prefix, cloudwatch_logs_default_retention, vpc, subnets, security_groups, environment, tags, find_address_timeout, skip_cleanup, aws_access_key_id, aws_secret_access_key, region_name, platform_version, fargate_use_private_ip, mount_points, volumes, mount_volumes_on_scheduler, **kwargs)
    786         self._lock = asyncio.Lock()
    787         self.session = get_session()
--> 788         super().__init__(**kwargs)
    789 
    790     def _client(self, name: str):

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/deploy/spec.py in __init__(self, workers, scheduler, worker, asynchronous, loop, security, silence_logs, name, shutdown_on_close, scheduler_sync_interval)
    258         if not self.asynchronous:
    259             self._loop_runner.start()
--> 260             self.sync(self._start)
    261             self.sync(self._correct_state)
    262 

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/utils.py in sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    308             return future
    309         else:
--> 310             return sync(
    311                 self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
    312             )

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    362     if error[0]:
    363         typ, exc, tb = error[0]
--> 364         raise exc.with_traceback(tb)
    365     else:
    366         return result[0]

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/utils.py in f()
    347             if callback_timeout is not None:
    348                 future = asyncio.wait_for(future, callback_timeout)
--> 349             result[0] = yield future
    350         except Exception:
    351             error[0] = sys.exc_info()

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/tornado/gen.py in run(self)
    760 
    761                     try:
--> 762                         value = future.result()
    763                     except Exception:
    764                         exc_info = sys.exc_info()

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/dask_cloudprovider/aws/ecs.py in _start(self)
   1001             "Hang tight! ",
   1002         ):
-> 1003             await super()._start()
   1004 
   1005     def _new_worker_name(self, worker_number):

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/deploy/spec.py in _start(self)
    287                 cls = import_term(cls)
    288             self.scheduler = cls(**self.scheduler_spec.get("options", {}))
--> 289             self.scheduler = await self.scheduler
    290         self.scheduler_comm = rpc(
    291             getattr(self.scheduler, "external_address", None) or self.scheduler.address,

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/dask_cloudprovider/aws/ecs.py in _()
    169             async with self.lock:
    170                 if not self.task:
--> 171                     await self.start()
    172                     assert self.task
    173             return self

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/dask_cloudprovider/aws/ecs.py in start(self)
    238     async def start(self):
    239         timeout = Timeout(60, "Unable to start %s after 60 seconds" % self.task_type)
--> 240         while timeout.run():
    241             try:
    242                 kwargs = self.task_kwargs.copy() if self.task_kwargs is not None else {}

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/dask_cloudprovider/utils/timeout.py in run(self)
     72                 return False
     73             else:
---> 74                 raise self.exception
     75         return True
     76 

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/dask_cloudprovider/aws/ecs.py in start(self)
    289 
    290                 async with self._client("ecs") as ecs:
--> 291                     response = await ecs.run_task(**kwargs)
    292 
    293                 if not response.get("tasks"):

/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/aiobotocore/client.py in _make_api_call(self, operation_name, api_params)
    178             error_code = parsed_response.get("Error", {}).get("Code")
    179             error_class = self.exceptions.from_code(error_code)
--> 180             raise error_class(parsed_response, operation_name)
    181         else:
    182             return parsed_response

InvalidParameterException: An error occurred (InvalidParameterException) when calling the RunTask operation: The specified capacity provider strategy cannot contain a capacity provider that is not associated with the cluster. Associate the capacity provider with the cluster or specify a valid capacity provider and try again.
distributed.deploy.cluster - WARNING - Failed to sync cluster info multiple times - perhaps there's a connection issue? Error:
Traceback (most recent call last):
  File "/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/comm/tcp.py", line 398, in connect
    stream = await self.client.connect(
  File "/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/tornado/tcpclient.py", line 275, in connect
    af, addr, stream = await connector.start(connect_timeout=timeout)
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/asyncio/tasks.py", line 492, in wait_for
    fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/comm/core.py", line 284, in connect
    comm = await asyncio.wait_for(
  File "/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/asyncio/tasks.py", line 494, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/deploy/cluster.py", line 131, in _sync_cluster_info
    await self.scheduler_comm.set_metadata(
  File "/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/core.py", line 808, in send_recv_from_rpc
    comm = await self.live_comm()
  File "/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/core.py", line 765, in live_comm
    comm = await connect(
  File "/home/ubuntu/miniconda3/envs/dask3/lib/python3.9/site-packages/distributed/comm/core.py", line 308, in connect
    raise OSError(
OSError: Timed out trying to connect to tcp://<redacted> after 30 s
```

</details>

**Comment by RichardScottOZ** [comment]
I just had this happen, too.

'''python
Exception ignored in: <function Cluster.__del__ at 0x000001DF84B68CA0>
Traceback (most recent call last):
  File "C:\Users\rscott\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\distributed\deploy\cluster.py", line 107, in __del__
    if self.status != Status.closed:
AttributeError: 'FargateCluster' object has no attribute 'status'
---------------------------------------------------------------------------
InvalidParameterException                 Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_17548/2949988005.py in <module>
      1 from dask_cloudprovider.aws import FargateCluster
----> 2 cluster = FargateCluster(image="867753693528.dkr.ecr.us-west-2.amazonaws.com/exploracorn-dataset:1", fargate_spot=True, aws_access_key_id=creds['AWS_ACCESS_KEY_ID'],aws_secret_access_key=creds['AWS_SECRET_ACCESS_KEY'], region_name='us-west-2')

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\dask_cloudprovider\aws\ecs.py in __init__(self, **kwargs)
   1473 
   1474     def __init__(self, **kwargs):
-> 1475         super().__init__(fargate_scheduler=True, fargate_workers=True, **kwargs)
   1476 
   1477 

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\dask_cloudprovider\aws\ecs.py in __init__(self, fargate_scheduler, fargate_workers, fargate_spot, image, scheduler_cpu, scheduler_mem, scheduler_timeout, scheduler_extra_args, scheduler_task_kwargs, scheduler_address, worker_cpu, worker_nthreads, worker_mem, worker_gpu, worker_extra_args, worker_task_kwargs, n_workers, workers_name_start, workers_name_step, cluster_arn, cluster_name_template, execution_role_arn, task_role_arn, task_role_policies, cloudwatch_logs_group, cloudwatch_logs_stream_prefix, cloudwatch_logs_default_retention, vpc, subnets, security_groups, environment, tags, find_address_timeout, skip_cleanup, aws_access_key_id, aws_secret_access_key, region_name, platform_version, fargate_use_private_ip, mount_points, volumes, mount_volumes_on_scheduler, **kwargs)
    786         self._lock = asyncio.Lock()
    787         self.session = get_session()
--> 788         super().__init__(**kwargs)
    789 
    790     def _client(self, name: str):

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\distributed\deploy\spec.py in __init__(self, workers, scheduler, worker, asynchronous, loop, security, silence_logs, name)
    274         if not self.asynchronous:
    275             self._loop_runner.start()
--> 276             self.sync(self._start)
    277             self.sync(self._correct_state)
    278 

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\distributed\deploy\cluster.py in sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    181             return future
    182         else:
--> 183             return sync(self.loop, func, *args, **kwargs)
    184 
    185     def _log(self, log):

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\distributed\utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    338     if error[0]:
    339         typ, exc, tb = error[0]
--> 340         raise exc.with_traceback(tb)
    341     else:
    342         return result[0]

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\distributed\utils.py in f()
    322             if callback_timeout is not None:
    323                 future = asyncio.wait_for(future, callback_timeout)
--> 324             result[0] = yield future
    325         except Exception as exc:
    326             error[0] = sys.exc_info()

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\tornado\gen.py in run(self)
    760 
    761                     try:
--> 762                         value = future.result()
    763                     except Exception:
    764                         exc_info = sys.exc_info()

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\dask_cloudprovider\aws\ecs.py in _start(self)
   1014             "Hang tight! ",
   1015         ):
-> 1016             await super()._start()
   1017 
   1018     def _new_worker_name(self, worker_number):

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\distributed\deploy\spec.py in _start(self)
    302 
    303         self.status = Status.starting
--> 304         self.scheduler = await self.scheduler
    305         self.scheduler_comm = rpc(
    306             getattr(self.scheduler, "external_address", None) or self.scheduler.address,

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\dask_cloudprovider\aws\ecs.py in _()
    169             async with self.lock:
    170                 if not self.task:
--> 171                     await self.start()
    172                     assert self.task
    173             return self

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\dask_cloudprovider\aws\ecs.py in start(self)
    238     async def start(self):
    239         timeout = Timeout(60, "Unable to start %s after 60 seconds" % self.task_type)
--> 240         while timeout.run():
    241             try:
    242                 kwargs = self.task_kwargs.copy() if self.task_kwargs is not None else {}

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\dask_cloudprovider\utils\timeout.py in run(self)
     72                 return False
     73             else:
---> 74                 raise self.exception
     75         return True
     76 

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\dask_cloudprovider\aws\ecs.py in start(self)
    289 
    290                 async with self._client("ecs") as ecs:
--> 291                     response = await ecs.run_task(**kwargs)
    292 
    293                 if not response.get("tasks"):

~\AppData\Local\Continuum\anaconda3\envs\pangeo3\lib\site-packages\aiobotocore\client.py in _make_api_call(self, operation_name, api_params)
    153             error_code = parsed_response.get("Error", {}).get("Code")
    154             error_class = self.exceptions.from_code(error_code)
--> 155             raise error_class(parsed_response, operation_name)
    156         else:
    157             return parsed_response

InvalidParameterException: An error occurred (InvalidParameterException) when calling the RunTask operation: The specified capacity provider strategy cannot contain a capacity provider that is not associated with the cluster. Associate the capacity provider with the cluster or specify a valid capacity provider and try again.
```

**Comment by davidvhill** [comment]
Hi @jacobtomlinson, 

I have this issue corrected by adding capacityProviders=['FARGATE', 'FARGATE_SPOT'] when creating the cluster.

Currently working through the contributing instructions now.

Dave

**Comment by jacobtomlinson** [comment]
Fixed in #348
